pennylane/_qubit_device.py

```diff
@@ -29,100 +30,164 @@
-        self._parameters.update(parameters)
+        self.check_validity(circuit.operations, circuit.observables)
 
         with self.execution_context():
-            self.pre_apply()
-
-            for operation in queue:
-                # Pass instances directly
-                self.apply(operation)
-
-            self.rotate_basis(observables)
-
-            self.post_apply()
+            # determine operations required to diagonalize observables
+            rotations = self.rotate_basis(circuit.observables)
 
-            self.pre_measure()
+            # apply all circuit operations
+            self.apply(circuit.operations, rotations)
 
-            results = self.statistics(observables)
+            # generate computational basis samples
+            self.generate_samples()
 
-            self.post_measure()
-
-            self._op_queue = None
-            self._obs_queue = None
-            self._parameters = None
+            # compute the required statistics
+            results = self.statistics(circuit.observables)
 
             # Ensures that a combination with sample does not put
             # expvals and vars in superfluous arrays
-            sample_return_types = (obs.return_type is Sample for obs in observables)
+            sample_return_types = (obs.return_type is Sample for obs in circuit.observables)
             if any(sample_return_types) and not all(sample_return_types):
                 return self._asarray(results, dtype="object")
 
             return self._asarray(results)
 
+    @abc.abstractmethod
+    def apply(self, operations, rotations=None, **kwargs):
+        """Apply quantum operations, and execute the quantum circuit.
+
+        Args:
+            operations (list[~.Operation]): operations to apply to the device
+            rotations (list[~.Operation]): operations that rotate the circuit
+                pre-measurement into the eigenbasis of the observables.
+        """
+
+    def rotate_basis(self, observables):
+        """Rotates the specified wires such that they
+        are in the eigenbasis of the provided observables.
+
+        Args:
+            observables (List[~.Observable]): the observables to diagonalize
+
+        Returns:
+            List[~.Operation]: the operations that diagonalize the observables
+        """
+        wires = []
+        rotation_gates = []
+
+        for observable in observables:
+            if hasattr(observable, "return_type") and observable.return_type == Sample:
+                self._memory = True  # make sure to return samples
+
+            rotation_gates.extend(observable.diagonalizing_gates())
+
+            for wire in observable.wires:
+                if isinstance(wire, int):
+                    wires.append(wire)
+                else:
+                    wires.extend(wire)
+
+        # Store the wires used by the observables such that
+        # an Identity is considered on the remaining wires
+        self._wires_used = set(wires)
```

Comment:
Looking now at the Forest plugin, I just realized, that the ``rotate_basis`` method will have to be called to populate ``_wires_used``, else it will just be an empty ``set`` and tests would fail.

Reply:
Should be fixed with the new `QubitDevice.active_wires` method!